### ISSUE TITLE
Add support for relative paths to stats.db

### ISSUE THREAD

**Issue by koldle** [mmatyas/pegasus-frontend#1097]
**Current behavior**: 
Currently, `stats.db` is being written using absolute paths. If the drive changes, the game records in the `plays` table become invalid. Additionally, due to the change in absolute paths, multiple records of the same game will be written in the `paths` table.

**Expected behavior**: 
When running Pegasus in portable mode, the `config` folder containing stats.db has a fixed path relative to pegasus.exe. It is possible to use relative paths to record game paths in the `paths` table. This way, even if the drive changes, the records in `stats.db` will remain valid.

*For scenarios where Pegasus is not running in portable mode, it is uncertain whether supporting relative paths would be meaningful. This aspect is open for discussion and further consideration.*

**Reason for expectation**:
`settings.txt` and `game_dirs.txt` are already support relative paths, adding support for relative paths to `stats.db` would significantly enhance portability and allow portable mode Pegasus to rely entirely on relative paths for configuration.

**Comment by WesleyFister** [comment]
This would be something very nice to have. Every so often I will move my Pegasus install to a different computer and have to manually edit the paths in stats.db to get it working. Also, Soon I'm going to have Pegasus on my desktop and Steam Deck, planning to use Syncthing to sync stats.db but this wouldn't work with absolute paths.

**Comment by Issac-v** [comment]
> This would be something very nice to have. Every so often I will move my Pegasus install to a different computer and have to manually edit the paths in stats.db to get it working. Also, Soon I'm going to have Pegasus on my desktop and Steam Deck, planning to use Syncthing to sync stats.db but this wouldn't work with absolute paths.

Yes, relative path can also make Peagasus fully adapt to multi-device synchronization, it'll be nice since we already have multiple platform clients.

**Comment by itzTheMeow** [comment]
A use case for non-portable systems is NixOS, where derivations are in the nix store at `/nix/store/xxxxxxxx-game/xxx` and can change if the system is rebuilt with new configuration.

**Comment by itzTheMeow** [comment]
I propose a game "id" system, a new option in the metadata files `id: string` which will be used in place of the game path in the stats.db if provided. This would allow games to have changing paths (such as my NixOS example or portable installations) and still have a static identifier for use in the database.

I noticed steam games (not sure about others) use `/home/$USER/steam:0000` as the path in the database. This could be changed to just the "steam:0000" path to make it static. I haven't tested any other providers, so im unsure what their IDs would look like.

The default full path can be used to keep backwards compatibility. The database could also automatically migrate itself:
In the case of non-metadata providers, they can be converted automatically (merge any `/.../.../steam:...` into just `steam:...` without issue.
For the metadata files, if a game has an ID added to it, the full path (`/nix/store/...` for example) could be merged into the game ID path.
This would ensure existing stats.db files don't break, and playtime is migrated automatically for users who want to use static IDs.

I would be willing to implement/document this myself (my C++ knowledge isn't amazing but i can figure it out) and submit a PR if this sounds like a decent idea. @mmatyas

**Comment by mmatyas** [comment]
I agree Steam games should just store `steam:xxxxx`, that's more like a bug than an intended feature. We should probably check if the "path" is actually a path, and not a provider-specific identifier.

I also agree absolute paths are not great, but that sounded like the least bad approach:

- Partial or full file hashes of the executable are not unique, nor any other binary info
- File system IDs don't work if the games are updated or moved between drives
- Relative paths don't work if either Pegasus or the games are moved
- Absolute paths at least work as long as the games do not move

I'm not a big fan of asking the users to maintain a unique ID for their games, but I agree that would be a working solution. In the metadata file, this could be a `unique-id: xxxxxx` field (probably with a value of string), and in the database, we could store `id:xxxxxx` elements.

I also expect several issues we should figure out:
- Duplicate IDs
- Duplicate IDs spanning multiple files
- ID format consistency (case sensitivity, whitespace), ie. are "skyrim_special_edition" and "Skyrim Special Edition" the same?
- "I have five thousand games, how do I assign unique IDs to them?"

If we introduce manual IDs, most people will likely want to use it, but I'm not sure we want to force people to maintain unique keys in their files.

**Comment by itzTheMeow** [comment]
I can't think of any other solutions at the moment that would avoid those first 4 points.

The IDs would be optional, the current functionality would stay as-is and people who want to opt-in to using IDs for more niche use cases (portable setup, nix, etc) can do so if they wish.

Regarding duplicate IDs, any of:
- Ignore both, print a warning, neither game gets the ID assigned.
- Ignore all but the first one, print a warning. (im not sure how deterministic the file reading is, this could create a race condition)
- Allow multiple games to "share" a "unique" ID. Still print a warning. This would mean both games share the playtime counter. (im not sure how this would work if both games are open at the same time. i suppose the latest playtime could be incremented to avoid this issue)

Regarding formatting:
Initially i thought maybe all lowercase, A-Z 0-9 -._ but im not sure that format is enforceable. I think it would be fine to just allow them to be any string, case sensitive. (`skyrim_special` is different from `skyrim special` which is also different from `Skyrim_special`)

Regarding a large game library:
I think if someone has a large set of games, they are using a provider (like steam) which wont be affected by this change (since their games arent file paths, at least in the case of steam and lutris), or they have scrapers/scripts to manage their library. In which case the scraper/script can be modified to add an ID if they wish.
If you already have a fairly large library of games, the absolute path issue probably does not apply to you.

----

> If we introduce manual IDs, most people will likely want to use it, but I'm not sure we want to force people to maintain unique keys in their files.

I was proposing this mostly as an optional alternative for more "technically inclined" people who *want* to use it for more niche use cases. If implemented, users could just ignore the option entirely and the frontend would work exactly the same as it did before with the absolute paths.

In my use case, i'm using nix scripts to [generate game metadata on the fly](https://github.com/itzTheMeow/dotfiles/blob/cffb79da84043ca310121374ac0d8c6693a8822e/nixos/gaming/hells-kitchen/default.nix) and this was the only game launcher I could find that actually lets you configure games 100% declaratively using text files.
If i modify the code that generates the binary for the game, it will have a new output hash and therefore a new file path, and playtime stats would never properly save for it.

**Comment by mmatyas** [comment]
Let's go with multiple games sharing the ID (ie. it's just a field for games), and case insensitive ID strings for ease of use (and sort of like Lutris' slug IDs). Otherwise sounds good 👍

**Comment by itzTheMeow** [comment]
It appears as though playtime is tracked for externally provided games (like steam/lutris) but is not displayed in the UI since it's never properly read.
```c
model::GameFile* const game_ptr = sctx.gamefile_by_filepath(path); // TODO: URI support
if (!game_ptr)
    continue;
```
There is a todo message to support URIs but it looks like support is nonexistent.
